File: task2/src/image_catch/scripts/image_store_node.py
#!/usr/bin/env python3
import rospy
import cv2
from sensor_msgs.msg import Image
from cv_bridge import CvBridge

class ImageProcessor:

    def __init__(self):
        rospy.init_node("image_store_node")
        rospy.loginfo("Node_init done")
        rospy.loginfo("Receiving")
        self.mg_sub = rospy.Subscriber("image_process",Image,self.img_callback,queue_size=1)

    def img_callback(self,msg):
        bridge = CvBridge()
        cv_image = bridge.imgmsg_to_cv2(msg,"passthrough")
        image_filename = "image_{}.png".format(str(rospy.Time.now()))
        image_path = '/home/ding/winter_train/ROS_vision/image_work/task2/image/'+image_filename
        cv2.imwrite(image_path,cv_image)
        rospy.logwarn("Received one")
        rospy.loginfo("Saved image {} to {}".format(image_filename, image_path))
        cv2.imshow(image_filename,cv_image)
        cv2.waitKey(2000)

if __name__ == "__main__":
    node = ImageProcessor()
    rospy.spin()

File: task2/src/image_catch/src/image_store_node.cpp
#include <ros/ros.h>
#include <sensor_msgs/Image.h>
#include <cv_bridge/cv_bridge.h>
#include <opencv2/opencv.hpp>

class ImageProcessor {
public:
    ImageProcessor() {
        ros::NodeHandle nh;
        ROS_INFO("Node_init done");
        ROS_INFO("Receiving");
        image_sub = nh.subscribe("image_process", 1, &ImageProcessor::imgCallback, this);
    }

    void imgCallback(const sensor_msgs::ImageConstPtr& msg) {
        cv_bridge::CvImagePtr cv_ptr;
        try {
            cv_ptr = cv_bridge::toCvCopy(msg, sensor_msgs::image_encodings::BGR8);
        } catch (cv_bridge::Exception& e) {
            ROS_ERROR("cv_bridge exception: %s", e.what());
            return;
        }

        std::string image_filename = "image_" + std::to_string(ros::Time::now().toSec()) + ".png";
        std::string image_path = "/home/ding/winter_train/ROS_vision/image_work/task2/image/" + image_filename;
        cv::imwrite(image_path, cv_ptr->image);
        ROS_WARN("Received one");
        ROS_INFO("Saved image %s to %s", image_filename.c_str(), image_path.c_str());
        cv::imshow(image_filename, cv_ptr->image);
        cv::waitKey(2000);
    }

private:
    ros::Subscriber image_sub;
};

int main(int argc, char** argv) {
    ros::init(argc, argv, "image_store_node");
    ImageProcessor processor;
    ros::spin();
    return 0;
}
